File: go/src/RumbleRacing/core/asset/txf/CLHE.go
package txf

import (
	"encoding/binary"
)

type CLHE struct {
	Entries []CLHEEntry
}

type CLHEEntry struct {
	CLDAStartOffset     uint32
	Unk2                uint16
	VRAM_Dest           uint16
	CLUTImageSizeLookup uint16
	PixelFormat         uint16
}

func parseCLHE(buf []byte) (*CLHE, error) {
	buf = buf[8:]
	// repeated 12 bytes
	var entries []CLHEEntry

	for i := 0; i+0xc <= len(buf); i += 0xc {
		data := buf[i : i+0xc]
		// fmt.Println(hex.Dump(data))
		entries = append(entries, CLHEEntry{
			CLDAStartOffset:     binary.LittleEndian.Uint32(data[0:4]),
			Unk2:                binary.LittleEndian.Uint16(data[0x4 : 0x4+2]),
			VRAM_Dest:           binary.LittleEndian.Uint16(data[0x6 : 0x6+2]),
			CLUTImageSizeLookup: binary.LittleEndian.Uint16(data[0x8 : 0x8+2]),
			PixelFormat:         binary.LittleEndian.Uint16(data[0xa : 0xa+2]),
		})
	}
	// fmt.Println(len(buf))

	return &CLHE{
		Entries: entries,
	}, nil
}

File: src/RumbleRacing/asset/txf/CLHE.ts
import ArrayBufferSlice from "../../../ArrayBufferSlice";
import { GSCLUTPixelStorageFormat } from "../../../Common/PS2/GS";

export interface CLHEEntry {
  cldaStartOffset: number;
  vramDest: number;
  clutImageSizeLookup: number;
  pixelFormat: GSCLUTPixelStorageFormat;
}

export interface CLHE {
  entries: CLHEEntry[];
}

export function parseCLHE(buf: ArrayBufferSlice): CLHE {
  const body = buf.subarray(8);
  const view = body.createDataView();
  const entries: CLHEEntry[] = [];

  for (let i = 0; i + 0xc <= body.byteLength; i += 0xc) {
    entries.push({
      cldaStartOffset: view.getUint32(i, true),
      vramDest: view.getUint16(i + 0x6, true),
      clutImageSizeLookup: view.getUint16(i + 0x8, true),
      pixelFormat: view.getUint16(i + 0xa, true),
    });
  }

  return { entries };
}
